packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/remote_constraint_provider.dart

```diff
@@ -17,40 +24,192 @@ abstract class RemoteLoggingConstraintProvider {
   LoggingConstraint? get loggingConstraint;
 }
 
+/// {@template aws_logging_cloudwatch.base_remote_constraints_provider}
+/// Base class for [RemoteLoggingConstraintProvider] to provide
+/// [LoggingConstraint] from a remote location and cache it.
+/// {@endtemplate}
+base class BaseRemoteLoggingConstraintProvider
+    with AWSDebuggable, AWSLoggerMixin
+    implements RemoteLoggingConstraintProvider, Closeable {
+  /// {@macro aws_logging_cloudwatch.base_remote_constraints_provider}
+  BaseRemoteLoggingConstraintProvider({
+    required DefaultRemoteConfiguration config,
+    FileStorage? fileStorage,
+  })  : _fileStorage = fileStorage,
+        _config = config,
+        _awsHttpClient = AWSHttpClient() {
+    _init();
+  }
+
+  /// A [BaseRemoteLoggingConstraintProvider] to use only for testing.
+  @protected
+  @visibleForTesting
+  BaseRemoteLoggingConstraintProvider.forTesting({
+    required DefaultRemoteConfiguration config,
+    required AWSHttpClient awsHttpClient,
+    FileStorage? fileStorage,
+  })  : _fileStorage = fileStorage,
+        _config = config,
+        _awsHttpClient = awsHttpClient {
+    _init();
+  }
+
+  final FileStorage? _fileStorage;
+
+  final DefaultRemoteConfiguration _config;
+
+  LoggingConstraint? _loggingConstraint;
+
+  final AWSHttpClient _awsHttpClient;
+
+  static const _cacheFileName = 'remoteloggingconstraints.json';
+
+  // The timer to refresh the constraint periodically.
+  Timer? _timer;
+
+  /// Retrives the runtime type name used for logging.
+  @override
+  String get runtimeTypeName => 'BaseRemoteConstraintsProvider';
+
+  /// Initializes the [BaseRemoteLoggingConstraintProvider] by fetching
+  /// the constraint from the endpoint initially and then
+  /// starting the refresh timer afterwards.
+  void _init() {
+    _readyCompleter.complete(_refreshConstraintPeriodically());
+  }
+
+  final Completer<void> _readyCompleter = Completer();
+
+  /// A future that completes when the [BaseRemoteLoggingConstraintProvider]
+  Future<void> get ready => _readyCompleter.future;
+
+  /// Creates a request to fetch the constraint from the endpoint.
+  @protected
+  Future<AWSBaseHttpRequest> createRequest() async {
+    final uri = Uri.parse(_config.endpoint);
+    return AWSHttpRequest(
+      method: AWSHttpMethod.get,
+      uri: uri,
+      headers: const {
+        AWSHeaders.accept: 'application/json; charset=utf-8',
+      },
+    );
+  }
+
+  /// Fetches the constraint from the endpoint and caches it.
+  Future<void> _fetchAndCacheConstraintFromEndpoint() async {
+    try {
+      final request = await createRequest();
+      final operation = _awsHttpClient.send(request);
+      final response = await operation.response;
+      final body = await response.decodeBody();
+      if (response.statusCode != 200) {
+        logger.error('Failed to fetch constraints', body);
```

Comment:
I would log the error code too. You can use a record to make it easy.

```suggestion
        logger.error('Failed to fetch constraints', (response.statusCode, body));
```